NetKVM/Common/ParaNdis_Protocol.cpp

```diff
@@ -913,6 +988,105 @@ void ParaNdis_ProtocolUnregisterAdapter(PARANDIS_ADAPTER *pContext, bool Unregis
     }
 }
 
+NTSTATUS CParaNdisProtocol::OnDeviceIoctl(PDEVICE_OBJECT DeviceObject, PIRP Irp)
+{
+    NTSTATUS status = STATUS_NOT_SUPPORTED;
+
+    PIO_STACK_LOCATION irpStack;
+    CParaNdisProtocol* prot = *(CParaNdisProtocol**)NdisGetDeviceReservedExtension(DeviceObject);
+    irpStack = IoGetCurrentIrpStackLocation(Irp);
+    ULONG code = irpStack->Parameters.DeviceIoControl.IoControlCode;
+    ULONG inSize = irpStack->Parameters.DeviceIoControl.InputBufferLength;
+    ULONG outSize = irpStack->Parameters.DeviceIoControl.OutputBufferLength;
+    PVOID buffer = Irp->AssociatedIrp.SystemBuffer;
+    TraceNoPrefix(0, "[%s] code %d, buf %p, in %d, out %d\n", __FUNCTION__, (code >> 2) & 0xfff, buffer, inSize, outSize);
+    Irp->IoStatus.Information = 0;
+    switch (code) {
+        case IOCTL_NETKVMD_QUERY_ADAPTERS:
+            status = prot->QueryAdapters(buffer, outSize, Irp->IoStatus.Information);
+            break;
+        case IOCTL_NETKVMD_SET_LINK:
+            status = prot->SetLink(buffer, inSize);
+            break;
+        default:
+            break;
+    }
+
+    Irp->IoStatus.Status = status;
+    if (status != STATUS_PENDING) {
+        IoCompleteRequest(Irp, IO_NO_INCREMENT);
+    }
+    else {
+        IoMarkIrpPending(Irp);
+    }
+    return status;
+}
+
+NTSTATUS CParaNdisProtocol::QueryAdapters(PVOID Buffer, ULONG Size, ULONG_PTR& Written)
+{
+    NETKVMD_ADAPTER* ad = (NETKVMD_ADAPTER*)Buffer;
+    NTSTATUS status = STATUS_SUCCESS;
+    Written = 0;
+
+    CMutexLockedContext protect(m_Mutex);
+
+    m_Adapters.ForEach([&](CAdapterEntry* e)
+    {
+        if (Size < sizeof(NETKVMD_ADAPTER)) {
+            status = STATUS_BUFFER_TOO_SMALL;
+            return false;
+        }
+        e->GetData(ad);
+        ad++;
+        Size--;
```

Comment:
Should be "- sizeof(NETKVMD_ADAPTER" and not "--"

Reply:
Fixed